### DISCUSSION TITLE
Is there any news on something like a wrapper for TensorFlow - Keras?

### DISCUSSION THREAD

**Discussion by tamoghnabh** [online-ml/river#639] [General]
If river can be made to work in conjunction with Keras, this library will become absolutely brilliant.

**Comment by raphaelsty** [comment]
Hi @Geronimo1995 ,

Could you provide more details? 

Personally, I use River with PyTorch on a daily basis to measure loss as I train my model. I also use River to calculate metrics in batchs mode. What's stopping us from using River for Keras now? We would be very happy to have your feedback.

**Reply by tamoghnabh** [reply]
Hi @raphaelsty 

I am in a self-conflicting situation as I write this. The main advantage of TensorFlow is the GPU computation which comes in handy when large datasets are available. River, as I understand does not need to be in that situation as the whole data is never available while streaming. Then again, if we do have a wrapper for Keras, many already developed Keras code blocks can be directly usable for streams of data. I am not sure if my comment is of any help to you or not. I would like to know more about how you are using River with PyTorch if that is possible in any way.

**Reply by raphaelsty** [reply]
River allows to update a model as soon as a new data is available. We therefore assume that we do not need GPUs because the real-time update of the model using CPU can absorb very large volumes of data. Also, River's philosophy is minimalist, the library is designed for Python data structures (dictionaries / pandas dataframe Work In Progress) and has very few dependencies. I don't really see River integrating natively into a Keras, TensorFlow or Pytorch pipeline because River serves different needs (Also technical difficulties related to the calculation of the gradient and the integration of the GPU is an important limitation). 

I see some limitations to the integration of these libraries in River:

- Creating a dependency with PyTorch / Keras / Tensorflow libraries is not possible. Installing all these libraries with River is too loud a requirement.

- River is optimized to process observations one by one and is faster than these libraries (CPU-based context). 

However, it would be interesting to target a specific Keras / Tensorflow / Pytorch method that is not available in river and would make sense for river. 

I use River on a daily basis to measure loss. With Pytorch I calculate the loss for each training batch and with `river.stats.Mean` I can get the average loss per epoch easily. (or running average loss using `river.stats.Rolling`). Also this week I used river to measure metrics related to the classification task (accuracy, recall...). It's not provided by PyTorch.
